internal/clientcache/internal/cache/refresh.go

```diff
@@ -131,6 +144,85 @@ func (r *RefreshService) refreshableUsers(ctx context.Context, in []*user) ([]*u
 	return ret, nil
 }
 
+// RefreshForSearch refreshes a specific resource type owned by the user associated
+// with the provided auth token id, if it hasn't been refreshed in a certain
+// amount of time. If the resources has been updated recently, or if it is known
+// that the user's boundary instance doesn't support partial refreshing, this
+// method returns without any requests to the boundary controller.
+// While the criteria of whether the user will refresh or not is almost the same
+// as the criteria used in Refresh, RefreshForSearch will not refresh any
+// data if there is not at least 1 refresh token stored for the resource. It
+// might make sense to change this in the future, but the reasoning used is
+// that we should not be making an initial load of all resources while blocking
+// a search query, in case we have not yet even attempted to load the resources
+// for this user yet.
+func (r *RefreshService) RefreshForSearch(ctx context.Context, authTokenid string, resourceType SearchableResource, opt ...Option) error {
+	const op = "cache.(RefreshService).RefreshForSearch"
+	if r.maxSearchRefreshTimeout > 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, r.maxSearchRefreshTimeout)
+		defer cancel()
+	}
+	at, err := r.repo.LookupToken(ctx, authTokenid)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	if at == nil {
+		return errors.New(ctx, errors.NotFound, op, "auth token not found")
+	}
+	u, err := r.repo.lookupUser(ctx, at.UserId)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+	if u == nil {
+		return errors.New(ctx, errors.NotFound, op, "user not found")
+	}
+	us, err := r.refreshableUsers(ctx, []*user{u})
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	switch len(us) {
+	case 0:
+		// The provided user is not refreshable, so do not refresh.
+		return nil
+	case 1:
+		u = us[0]
+	default:
+		return errors.New(ctx, errors.Internal, op, "unexpected number of refreshable users")
+	}
+
+	tokens, err := r.cleanAndPickAuthTokens(ctx, u)
+	if err != nil {
+		return errors.Wrap(ctx, err, op)
+	}
+
+	switch resourceType {
+	case Targets:
+		rtv, err := r.repo.lookupRefreshToken(ctx, u, targetResourceType)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		if rtv != nil && time.Since(rtv.UpdateTime) > r.maxSearchStaleness {
+			if err := r.repo.refreshTargets(ctx, u, tokens, opt...); err != nil {
+				return errors.Wrap(ctx, err, op)
+			}
+		}
+	case Sessions:
+		rtv, err := r.repo.lookupRefreshToken(ctx, u, sessionResourceType)
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		if rtv != nil && time.Since(rtv.UpdateTime) > r.maxSearchStaleness {
+			if err := r.repo.refreshSessions(ctx, u, tokens, opt...); err != nil {
+				return errors.Wrap(ctx, err, op)
+			}
+		}
+	}
```

Comment:
Can we have a default case with an error so we know if we forget to add new resources here?